File: Java/projects/BankAccount/BankAccount.java
package projects.BankAccount;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import projects.BankAccount.Utils.*;
class BankAccount {
    private static class Account {
        String username;
        String passhash;
        double balance;

        Account(String username, String passhash, double balance) {
            this.username = username;
            this.passhash = passhash;
            this.balance = balance;
        }
    }

    private static List<Account> accounts = new ArrayList<>();
    private static final double WITHDRAW_LIMIT = 3000;
    private String authUser = null;

    static {
        // passwords shown in plaintext for demonstration purposes
        accounts.add(new Account("user1", hashPw("password"), 1.0));
        accounts.add(new Account("user2", hashPw("123456"), 2.0));
        accounts.add(new Account("user3", hashPw("verysecure"), 3.0));
    }

    public BankAccount(String name, String password) {
        for (Account account : accounts) {
            if (account.username.equals(name) && account.passhash.equals(hashPw(password))) {
                Utils.printInBox("Welcome, " + name);
                this.authUser = name;
                return;
            }
        }
        Utils.printInBox(Utils.formatAnsi("Bad username or password.", "red", true));
    }

    public Double getBalance() {
        if (authUser == null) {
            Utils.printInBox(Utils.formatAnsi("No authenticated user.", "red", true));
            return null;
        }
        for (Account account : accounts) {
            if (account.username.equals(authUser)) {
                return account.balance;
            }
        }
        return null;
    }

    public void withdraw(double amount) {
        if (authUser == null) {
            Utils.printInBox(Utils.formatAnsi("No authenticated user.", "red", true));
            return;
        }
        if (amount > WITHDRAW_LIMIT) {
            Utils.printInBox(Utils.formatAnsi(amount + " is above the withdraw limit." + "(limit: " + WITHDRAW_LIMIT + ")" , "red", true));
            return;
        }
        for (Account account : accounts) {
            if (account.username.equals(authUser)) {
                if (account.balance - amount < 0) {
                    Utils.printInBox("You don't have that much money you brokie.\n Balance: " + account.balance + "\n Attempted withdrawal: " + amount);
                } else {
                    account.balance -= amount;
                    Utils.printInBox(Utils.formatAnsi("New Balance: ", "green", true) + account.balance);
                }
                return;
            }
        }
    }

    public void deposit(double amount) {
        if (authUser == null) {
            Utils.printInBox(Utils.formatAnsi("No authenticated user.", "red", true));
            return;
        }
        for (Account account : accounts) {
            if (account.username.equals(authUser)) {
                account.balance += amount;
                Utils.printInBox(Utils.formatAnsi("New Balance: ", "green", true) + account.balance);
                return;
            }
        }
    }

    private static String hashPw(String pw) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(pw.getBytes());
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                hexString.append(String.format("%02x", b));
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("Error hashing password", e);
        }
    }
}

File: Python/examples/atm/BankAccount.py
import hashlib

def hash_pw(pw):
    return hashlib.sha256(pw.encode()).hexdigest()

class BankAccount():
    accounts = [ 
        {
            "username": "user1",
            "passhash": "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8", # "password"
            "balance": "1"
        },
         {
            "username": "user2",
            "passhash": "8d969eef6ecad3c29a3a629280e686cf0c3f5d5a86aff3ca12020c923adc6c92", # "123456"
            "balance": "2"
        },
        {
            "username": "user3",
            "passhash": "6ddbb91a63b88bad59b108fa7a4b236155f5b1b1dd2dc9580798624dcbc2dcee", # "verysecure"
            "balance": "3"
        },
        
    ]
    auth = None
    withdraw_limit = 9999
    def __init__(self, name, password):
        if self.auth != None:
            self.auth = None
        for account in self.accounts:
            if (account["username"] == name):
                if (account["passhash"] == hash_pw(password)):
                    print(f'Welcome, {name}')
                    self.auth = name
                else:
                    print(f'wrong password for user: {name}')
    def getBalance(self):
        if self.auth == None:
            print(f'Error: no authenticated user.')
        else:
            return next((account["balance"] for account in self.accounts if account["username"] == self.auth), None)
    def withdraw(self, amount):
        if self.auth == None:
            print(f'Error: no authenticated user.')
        elif amount > self.withdraw_limit:
            print(f'Error: {amount} is above the withdraw limit')
        else:
            account = next((account for account in self.accounts if account["username"] == self.auth), None)
            bal = float(account['balance'])
            if (bal - float(amount)) < 0:
                print(f"You don't have that much money you brokie.\n Balance: {account['balance']}\n Attempted withdrawal: {amount}")
            else:
                account["balance"] -= amount
                print(f"New Balance: {account['balance']}")
    def deposit(self, amount):
        if self.auth == None:
            print(f'Error: no authenticated user.')
        else:
            account = next((account for account in self.accounts if account["username"] == self.auth), None)
            account["balance"] = float(account["balance"]) + amount
            print(f"New Balance: {account['balance']}")
        
                
        

